README
README for morph
================

> **NOTA BENE:** This document is very much work-in-progress, and anything
> and everything may and will change at little or no notice. If you see
> problems, mail baserock-dev@baserock.org.

`morph` builds binaries for [Baserock](http://www.baserock.org/),
an appliance Linux solution. Please see the website for overall information.


Usage
-----

The Baserock builds are controlled by **morphology** files,
which are build recipes. See below for their syntax. Everything
in Baserock is built from git commits.
Morphologies must be committed in git before building. The `morph` tool is
used to actually run the build. The usual workflow is this:

* put the morphology for an upstream project with its source code
* put other morphologies in the `morphs` (note plural) repository
* run `morph` to build stuff

`morph --help` will provide some information, though a full guide is
really required. Meanwhile a short usage to build a disk image:

    morph init workspace
    cd workspace
    morph checkout baserock:baserock/definitions master
    cd master/baserock/baserock/definitions
    morph build base-system-x86_64-generic

For deploying you need to create a cluster morphology. Here is an
example to deploy to a raw disk image.

    name: foo
    kind: cluster
    systems:
    - morph: base-system-x86_64-generic
      repo: baserock:baserock/definitions
      ref: master
      deploy:
          my-raw-disk-image:
              type: rawdisk
              location: /src/tmp/testdev.img
              DISK_SIZE: 4G

To deploy it, you only need to run `morph deploy` with the cluster morphology
created:

    morph deploy foo

You can write a configuration file to avoid having to write options on
the command line every time. Put it in `~/.morph.conf` and make it look
something like this:

    [config]
    cachedir = /home/username/baserock/cache
    log = /home/username/baserock/morph.log
    log-max = 200M
    trove-host = git.baserock.org

All of the above settings apart from `log` are the defaults, so may be omitted.


Morphology file syntax
----------------------

YAML is used for the morphology syntax. For example, to build a chunk:

    name: foo
    kind: chunk
    configure-commands:
    - ./configure --prefix="$PREFIX"
    build-commands:
    - make
    test-commands:
    - make check
    install-commands:
    - make DESTDIR="$DESTDIR" install

For all morphologies, use the following fields:

* `name`: the name of the morphology; it must currently match the filename
  (without the `.morph` suffix); **required**
* `kind`: the kind of thing being built; **required**

For chunks, use the following fields:


* `build-system`: if the program is built using a build system known to
  `morph`, you can set this field and avoid having to set the various
  `*-commands` fields; the commands that the build system specifies can
  be overridden; the following build-systems are known:

  - `autotools`
  - `python-distutils`
  - `cpan`
  - `cmake`
  - `qmake`

  optional

* `pre-configure-commands`: a list of shell commands to run at
  the configuration phase of a build, before the list in `configure-commands`;
  optional
* `configure-commands`: a list of shell commands to run at the configuraiton
  phase of a build; optional
* `post-configure-commands`: a list of shell commands to run at
  the configuration phase of a build, after the list in `configure-commands`;
  optional

* `pre-build-commands`: a list of shell commands to run at
  the build phase of a build, before the list in `build-commands`;
  optional
* `build-commands`: a list of shell commands to run to build (compile) the
  project; optional
* `post-build-commands`: a list of shell commands to run at
  the build phase of a build, after the list in `build-commands`;
  optional

* `pre-test-commands`: a list of shell commands to run at
  the test phase of a build, before the list in `test-commands`;
  optional
* `test-commands`: a list of shell commands to run unit tests and other
  non-interactive tests on the built but un-installed project; optional
* `post-test-commands`: a list of shell commands to run at
  the test phase of a build, after the list in `test-commands`;
  optional

* `pre-install-commands`: a list of shell commands to run at
  the install phase of a build, before the list in `install-commands`;
  optional
* `install-commands`: a list of shell commands to install the built project;
  the install should go into the directory named in the `DESTDIR` environment
  variable, not the actual system; optional
* `post-install-commands`: a list of shell commands to run at
  the install phase of a build, after the list in `install-commands`;
  optional

* `max-jobs`: a string to be given to `make` as the argument to the `-j`
  option to specify the maximum number of parallel jobs; the only sensible
  value is `"1"` (including the quotes), to prevent parallel jobs to run
  at all; parallel jobs are only used during the `build-commands` phase,
  since the other phases are often not safe when run in parallel; `morph`
  picks a default value based on the number of CPUs on the host system;
  optional

* `chunks`: a key/value map of lists of regular expressions;
  the key is the name
  of a binary chunk, the regexps match the pathnames that will be
  included in that chunk; the patterns match the pathnames that get installed
  by `install-commands` (the whole path below `DESTDIR`); every file must
  be matched by at least one pattern; by default, a single chunk gets
  created, named according to the morphology, and containing all files;
  optional

For strata, use the following fields:

* `build-depends`: a list of strings, each of which refers to another
  stratum that the current stratum depends on. This list may be omitted
  or empty if the stratum does not depend on anything else.
* `chunks`: a list of key/value mappings, where each mapping corresponds
  to a chunk to be included in the stratum; the mappings may use the
  following keys: `name` is the chunk's name (may be different from the
  morphology name), `repo` is the repository in which to find (defaults to
  chunk name), `ref` identifies the commit to use (typically a branch
  name, but any tree-ish git accepts is ok), and `morph` is the name
  of the morphology to use and is optional. In addition to these keys,
  each of the sources MUST specify a list of build dependencies using the
  `build-depends` field. This field may be omitted to make the source
  depend on all other chunks that are listed earlier in the `chunks`
  list. The field may be an empty list to indicate that the chunk does
  not depend on anything else in the same stratum. To specify one or
  more chunk dependencies, `build-depends` needs to be set to a list
  that contains the names of chunks that the source depends on in the
  same stratum. These names correspond to the values of the `name`
  fields of the other chunks.

For systems, use the following fields:

* `strata`: a list of names of strata to be included in the system. Unlike
  chunks, the stratum morphs must all be in the same Git repository as the
  system morphology. The value of the `morph` field will be taken as the
  artifact name; if this causes ambiguity then an `alias` may be specified as
  well. **required**

Example chunk (simplified commands):

    name: eglibc
    kind: chunk
    configure-commands:
    - mkdir o
    - cd o && ../libc/configure --prefix=/usr
    build-commands:
    - cd o && make
    install-commands:
    - cd o && make install_root="$DESTDIR" install

Example stratum:

    name: foundation
    kind: stratum
    chunks:
    - name: fhs-dirs
      repo: upstream:fhs-dirs
      ref: baserock/bootstrap
      build-depends: []
    - name: linux-api-headers
      repo: upstream:linux
      ref: baserock/morph
      build-depends:
      - fhs-dirs
    - name: eglibc
      repo: upstream:eglibc
      ref: baserock/bootstrap
      build-depends:
      - linux-api-headers
    - name: busybox
      repo: upstream:busybox
      ref: baserock/bootstrap
      build-depends:
      - fhs-dirs
      - linux-api-headers

Example system:

    name: base
    kind: system
    strata:
    - morph: foundation
    - morph: linux-stratum

Note that currently, unknown keys in morphologies are silently ignored.


Build environment
-----------------

When `morph` runs build commands, it clears the environment of all
variables and creates new ones. This is so that the build will run
more consistently across machines and developers.

See the `morphlib/buildenvironment.py` file for details on what
environment variables are set.

Morph also constructs a staging area for every build, composed of its
build-dependencies, so everything that is used for a build is traceable
and reproducible.


Hacking morph
-------------

When running Morph from a Git checkout, remember to set PYTHONPATH to
point to your checkout. This will cause Morph to load the plugins and
write extensions from your checkout correctly.

Run the test suite with this command:

    ./check --full

If your /tmp is a tmpfs you may need to set TMPDIR to a different path,
as there are tests for large disk image deploys.

Install CoverageTestRunner (from <http://liw.fi/coverage-test-runner/>),
and check out the `cmdtest` utility (from <http://liw.fi/cmdtest/>).

Run the checks before submitting a patch, please.


Legalese
--------

Copyright (C) 2011-2014  Codethink Limited

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.



NEWS
NEWS for Morph
==============

This file contains high-level summaries of user-visible changes in
each Morph release.

Version 14.26, released 2014-06-27
----------------------------------

* Smarter git caching behaviour, cached repos will only be updated when
  necessary
* Morph deploy now lets the user specify which systems they want to deploy
  within a cluster
* Various bug fixes

Version 14.24, released 2014-06-13
----------------------------------

* rawdisk deployments check that they have the btrfs module loaded first
* distbuild should busy-wait less
* fetching of artifacts should be atomic, so a failure to fetch the
  metadata of an artifact doesn't confuse the build system when we have
  the chunk, but no metadata
* `morph deploy` now defaults to `--no-git-update`
* `morph gc` now cleans up failed deployments, so they aren't left around
  if morph terminates uncleanly
* `morph edit` now only takes the name of the chunk, rather than the
  name of the system and stratum that chunk is in

Version 14.23, released 2014-06-06
----------------------------------

New feature:

* Initramfs support

There have also been a number of fixes to distbuild, and the
`morph copy-artifacts` command has been replaced by `morph list-artifacts`.

Version 14.22, released 2014-05-29
----------------------------------

New features:

* VirtualBox deployment now supports Vagrant. See:
  <http://wiki.baserock.org/guides/vagrant-basebox/>

* Additional checks when deploying upgrades with the 'ssh-rsync' extension.

Additional bug fixes described in the git log.

Version 14.20, released 2014-05-14
----------------------------------

New features include:

* New CPU architecture: armv7lhf (ARM hard float).

* Artifact splitting for chunk and stratum artifacts.

* Components that can be used to set up a distributed build network of Morph
  build workers.

* Built-in documentation for some extensions, see `morph help-extensions` and
  `morph help <extension>`.

* Nested deployment by `morph deploy`.

* Support for adding binaries to Git repos when used with Trove. See the `morph
  add-binary` and `morph push` commands.

Many additional changes are described in the Git log.

Version 13, released 2014-01-10
-------------------------------

New features added:

* New CPU architecture: ppc64 architecture (POWER PC 64-bit). This is
  the change specific for Morph. There are changes to the morphologies
  (in a different git repository) to actually build such systems.

* `morph build` and `morph deploy` now allow `.morph` suffixes in
  command line arguments. The suffixes are stripped internally, so
  Morph behaves as if they suffix wasn't there in the first place.

* The `morph build` command is now a new implementation. The old
  implementation is still available as `morph old-build`, just in case
  the new code is buggy, but will be removed in a future release.
  Likewise, `morph deploy` has a new implementation, but no
  `old-deploy`. Both new implementations should work exactly as the
  old ones, except for bugs.

Bugs fixed:

* When Morph reads git configuration files, it now correctly handles
  whitespace at the end of configuration values.

* `morph deploy` no longer creates and pushes a temporary build
  branch. Pushing it wasn't useful, merely wasteful.

* `morph deploy` now allows cross-architecture deployments, and
  and `morph cross-bootstrap` checks that the system is being built
  supports the target architecture.

Other user-visible changes:

* When preparing to build (when construcing the build graph), Morph
  now reports the ref (SHA1) it uses for each stratum.

* Systems being built must now have at least one stratum, and the
  strata in a system must have at least one chunk that is built using
  the normal (staging area) mode, rather than bootstrap mode.

Version 12, released 2013-11-15
-------------------------------

* NEWS file added.


morphlib/morph2.py
# Copyright (C) 2012-2014  Codethink Limited
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


import re

import morphlib
from morphlib.util import OrderedDict, json

class Morphology(object):

    '''An in-memory representation of a morphology.

    This is a parsed version of the morphology, with rules for default
    values applied. No other processing.

    '''

    static_defaults = {
        'chunk': [
            ('description', ''),
            ('pre-configure-commands', None),
            ('configure-commands', None),
            ('post-configure-commands', None),
            ('pre-build-commands', None),
            ('build-commands', None),
            ('post-build-commands', None),
            ('pre-test-commands', None),
            ('test-commands', None),
            ('post-test-commands', None),
            ('pre-install-commands', None),
            ('install-commands', None),
            ('post-install-commands', None),
            ('devices', None),
            ('products', []),
            ('max-jobs', None),
            ('build-system', 'manual')
        ],
        'stratum': [
            ('chunks', []),
            ('description', ''),
            ('build-depends', None),
        ],
        'system': [
            ('strata', []),
            ('description', ''),
            ('arch', None),
            ('configuration-extensions', []),
        ],
        'cluster': [
            ('description', ''),
        ],
    }

    @staticmethod
    def _load_json(text):
        return json.loads(text, object_pairs_hook=OrderedDict,
                          encoding='unicode-escape')

    @staticmethod
    def _dump_json(obj, f):
        text = json.dumps(obj, indent=4, encoding='unicode-escape')
        text = re.sub(" \n", "\n", text)
        f.write(text)
        f.write('\n')

    def __init__(self, text):
        self._dict, self._dumper = self._load_morphology_dict(text)
        self._set_defaults()
        self._validate_children()

    def __getitem__(self, key):
        return self._dict[key]

    def __contains__(self, key):
        return key in self._dict

    # Not covered by tests, since it's trivial, morph2 is going away
    # and only exists so the new morphology validation code can use it.
    def get(self, key, default=None): # pragma: no cover
        try:
            return self[key]
        except KeyError:
            return default

    def get_commands(self, which):
        '''Return the commands to run from a morphology or the build system'''
        if self[which] is None:
            attr = '_'.join(which.split('-'))
            bs = morphlib.buildsystem.lookup_build_system(self['build-system'])
            return getattr(bs, attr)
        else:
            return self[which]

    def keys(self):
        return self._dict.keys()

    def _load_morphology_dict(self, text):
        '''Load morphology, identifying whether it is JSON or YAML'''

        try:
            data = self._load_json(text)
            dumper = self._dump_json
        except ValueError as e:  # pragma: no cover
            data = morphlib.yamlparse.load(text)
            dumper = morphlib.yamlparse.dump

        if data is None:
            raise morphlib.YAMLError("Morphology is empty")
        if type(data) not in [dict, OrderedDict]:
            raise morphlib.YAMLError("Morphology did not parse as a dict")

        return data, dumper

    def _validate_children(self):
        if self['kind'] == 'system':
            names = set()
            for info in self['strata']:
                name = info.get('alias', info['morph'])
                if name in names:
                   raise ValueError('Duplicate stratum "%s"' % name)
                names.add(name)
        elif self['kind'] == 'stratum':
            names = set()
            for info in self['chunks']:
                name = info.get('alias', info['name'])
                if name in names:
                   raise ValueError('Duplicate chunk "%s"' % name)
                names.add(name)
        elif self['kind'] == 'cluster':
             if not 'systems' in self:
                 raise KeyError('"systems" not found')
             if not self['systems']:
                 raise ValueError('"systems" is empty')
             for system in self['systems']:
                 if 'morph' not in system:
                     raise KeyError('"morph" not found')
                 if 'deploy-defaults' in system:
                     if not isinstance(system['deploy-defaults'], dict):
                         raise ValueError('deploy defaults for morph "%s" '
                                          'are not a mapping: %r'
                                          % (system['morph'],
                                             system['deploy-defaults']))
                 if 'deploy' in system:
                     for system_id, deploy_params in system['deploy'].items():
                         if not isinstance(deploy_params, dict):
                             raise ValueError('deployment parameters for '
                                              'system "%s" are not a mapping:'
                                              ' %r'
                                              % (system_id, deploy_params))

    def _set_default_value(self, target_dict, key, value):
        '''Change a value in the in-memory representation of the morphology

        Record the default value separately, so that when writing out the
        morphology we can determine whether the change from the on-disk value
        was done at load time, or later on (we want to only write back out
        the later, deliberate changes).

        '''
        target_dict[key] = value
        target_dict['_orig_' + key] = value

    def _set_defaults(self):
        if 'max-jobs' in self:
            self._set_default_value(self._dict, 'max-jobs',
                                    int(self['max-jobs']))

        for name, value in self.static_defaults[self['kind']]:
            if name not in self._dict:
                self._set_default_value(self._dict, name, value)

        if self['kind'] == 'stratum':
            self._set_stratum_defaults()
        elif self['kind'] == 'cluster':
            self._set_cluster_defaults()

    def _set_stratum_defaults(self):
        for source in self['chunks']:
            if 'repo' not in source:
                self._set_default_value(source, 'repo', source['name'])
            if 'morph' not in source:
                self._set_default_value(source, 'morph', source['name'])
            if 'build-depends' not in source:
                self._set_default_value(source, 'build-depends', None)
            if 'build-mode' not in source:
                self._set_default_value(source, 'build-mode', 'staging')
            if 'prefix' not in source:
                self._set_default_value(source, 'prefix', '/usr')

    def _set_cluster_defaults(self):
        if 'systems' in self and self['systems']:
            for system in self['systems']:
                if 'deploy-defaults' not in system:
                    self._set_default_value(system,
                                            'deploy-defaults',
                                            dict())
                if 'deploy' not in system:
                    self._set_default_value(system,
                                            'deploy',
                                            dict())

    def lookup_child_by_name(self, name):
        '''Find child reference by its name.

        This lookup honors aliases.

        '''

        if self['kind'] == 'system':
            for info in self['strata']:
                source_name = info.get('alias', info['morph'])
                if source_name == name:
                    return info
        elif self['kind'] == 'stratum':
            for info in self['chunks']:
                source_name = info.get('alias', info['morph'])
                if source_name == name:
                    return info
        raise KeyError('"%s" not found' % name)

    def _apply_changes(self, live_dict, original_dict):
        '''Returns a new dict updated with changes from the in-memory object

        This allows us to write out a morphology including only the changes
        that were done after the morphology was loaded -- not the changes done
        to set default values during construction.

        '''
        output_dict = {}

        for key in live_dict.keys():
            if key.startswith('_orig_'):
                continue

            value = self._apply_changes_for_key(key, live_dict, original_dict)
            # VILE HACK to preserve nulls in repo/ref fields
            if value is not None or key in ('repo', 'ref'):
                output_dict[key] = value
        return output_dict

    def _apply_changes_for_key(self, key, live_dict, original_dict):
        '''Return value to write out for one key, recursing if necessary'''

        live_value = live_dict.get(key, None)
        orig_value = original_dict.get(key, None)

        if type(live_value) in [dict, OrderedDict] and orig_value is not None:
            # Recursively apply changes for dict
            result = self._apply_changes(live_value, orig_value)
        elif type(live_value) is list and orig_value is not None:
            # Recursively apply changes for list (existing, then new items).
            result = []
            for i in range(0, min(len(orig_value), len(live_value))):
                if type(live_value[i]) in [dict, OrderedDict]:
                    item = self._apply_changes(live_value[i], orig_value[i])
                else:
                    item = live_value[i]
                result.append(item)
            for i in range(len(orig_value), len(live_value)):
                if type(live_value[i]) in [dict, OrderedDict]:
                    item = self._apply_changes(live_value[i], {})
                else:
                    item = live_value[i]
                result.append(item)
        else:
            # Simple values. Use original value unless it has been changed from
            # the default in memmory.
            if live_dict[key] == live_dict.get('_orig_' + key, None):
                result = original_dict.get(key, None)
            else:
                result = live_dict[key]
        return result

    def update_text(self, text, output_fd, convert_to=None):
        '''Write out in-memory changes to loaded morphology text

        Similar in function to update_file().

        '''
        original_dict, dumper = self._load_morphology_dict(text)

        if convert_to == 'json': # pragma: no cover
            dumper = self._dump_json
        elif convert_to == 'yaml': # pragma: no cover
            dumper = morphlib.yamlparse.dump

        output_dict = self._apply_changes(self._dict, original_dict)
        dumper(output_dict, output_fd)

    def update_file(self, filename, output_fd=None, **kws): # pragma: no cover
        '''Write out in-memory changes to on-disk morphology file

        This function reads the original morphology text from 'filename', so
        that it can avoid writing out properties that are set in memory
        to their default value but weren't specified by the user at all.

        '''
        with open(filename, 'r') as f:
            text = f.read()

        with output_fd or morphlib.savefile.SaveFile(filename, 'w') as f:
            self.update_text(text, f, **kws)
